### ISSUE TITLE
"Request method 'POST' not supported" error when streaming a large dataset

### ISSUE THREAD

**Issue by jsec516** [MadKudu/node-marketo#36]
I was trying to import leads from marketo using streams (it was almost 200K data). After importing (~100K) I got consistent error

code was something like 

``````
       ```
       // Only pass to Marketo API a subset of the options passed to this function.
        var opts = {
          listId: options.listId,
          batchSize: options.batchSize || 300,
          fields: options.fields
        };
        // Return a stream.
        return client.list.getLeads(options.listId, opts).then(function(res){
          return M.streamify(Promise.resolve(res));
        });
    ```
``````

Error is:

`{"requestId":"1751d#1570622065b","success":false,"errors":[{"code":"605","message":"Request method 'POST' not supported"}],"level":"error","message":"","timestamp":"2016-09-07T19:31:16.033Z"}`

as far as I understand stream is stopped after error happens and there is no way to get the remains data. Is there any way to resume the stream to get remain leads from marketo.

**Comment by pcothenet** [comment]
@jsec516 we're experiencing the same error on our end.

I haven't been able to understand why this error comes up in the first place, do you have any idea?

**Comment by peaches** [comment]
@jsec516, is it always the same number of leads that you receive this error?

A few thoughts:

- The lead API is implemented as a POST with the `_method` parameter of `GET`. Could this be causing that error?
- Maybe it's a one time error on marketo's side and the code should check for 605 (after a pagination) and retry a few times.
- Add logging for previous response metadata to see if something's weird, and hopefully be able to reuse the `nextPageToken` to manually inspect.

Good luck!

**Comment by jsec516** [comment]
@pcothenet No it's not. It happens for different number of leads.
@peaches I'm not sure actually, but when it happens when I've already hit marketo api few times (suppose I've already imported ~100k). Finally it worked for me (I've been able to get a list of lead on redis  which contains almost 175k without error). I've was using stream for that (`streamify` method) though I've to  tweaks little bit. In my response `moreResult` property wasn't available (though the code expects it should be for lazy loading next page data), so I've made little changes to work it for me (you can check my fork copy if you want https://github.com/jsec516/node-marketo).

If you think it's useful I can submit a PR for that.

Thanks for your reply.

**Comment by ghiden** [comment]
This error happens very randomly and the size of data does not matter.
I still don't know what causes this error but changing method from POST to GET does eliminate the issue.
So, I think the method should check parameter size and if it exceeds 8K, then it should switch to POST.

**Comment by pcothenet** [comment]
OK, I kept having this error, so I switch the call to a full GET in #49 (in 0.4.0). I will confirm if that fixes our issue.

**Comment by ghiden** [comment]
@pcothenet I switched to GET on my own fork last year in May, and it's been working fine.

**Comment by pcothenet** [comment]
Thanks! Seems to be working fine in test.
I inherited this project, so not sure why the previous GET/POST had been
implemented.  





On Mon, Feb 26, 2018 1:08 PM, Hidenari Nozaki notifications@github.com  wrote:
@pcothenet  I switched to GET on my own fork last year in May, and it's been
working fine.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.



Paul CothenetCo-founder & CTO |MadKudu
